[modbus_controller] fix #6998

gotnone opened the pull request:
`server_register` address ranges may not overlap

# What does this implement/fix?
This fixes https://github.com/esphome/issues/issues/6998

Additional validator `validate_address` added to `modbus_controller/__init__.py` 
- Check for overlapping address ranges in the the list of `server_registers`
- Calculate `last_address` based on `address` and `value_type` fields
- Check if `last_address` is `cv.uint16_t`
- Modify `address` verification to use `cv.uint16_t` instead of `cv.positive_int`

>[!NOTE]
> This PR may break user configurations that rely on the current quirk where the address ranges are permitted to overlap.

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Potentially Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes https://github.com/esphome/issues/issues/6998

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4889

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
modbus_controller:
  - modbus_id: modbus_server
    address: 0x1
    server_registers:
      - address: 0x0000
        value_type: U_QWORD
        read_lambda: |-
          return(22);
      - address: 0x0001
        value_type: U_WORD
        read_lambda: |-
          return(42);
```
## Output for correctly detected overlapping addresses

```
INFO ESPHome 2025.5.0-dev
INFO Reading configuration breakmodbus.yaml...
Failed config

modbus_controller: [source breakmodbus.yaml:50]
  - modbus_id: modbus_server
    address: 1
    server_registers:
      - address: 0
        value_type: U_QWORD
        return(22);

      Address range overlap for 1..1.
      - address: 1
        value_type: U_WORD
        return(42);
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

 I don't know how to test to prove it prevents overlapping addresses, besides examining the results of a good configuration vs a configuration file with overlapping addresses

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Added note specifying that `server_register` address ranges may not overlap

Commit:
[modbus_controller] fix #6698
`server_register` address ranges may not overlap

Commit:
Merge remote-tracking branch 'origin/dev' into fix_6698

jesserockz commented on esphome/components/modbus_controller/__init__.py:
```diff
@@ -126,13 +126,43 @@
+    return validator
+
+
+ModbusServerRegisterSet = set()
```
Please use `CORE.data[DOMAIN]["some key"]` for storing things like this.
Will need to add `DOMAIN = "modbus_controller` to the top of the file too

gotnone replied:
OK.  I think I have this correct, now.

gotnone replied:
I think I have this fixed like in the ble_server example

jesserockz commented on esphome/components/modbus_controller/__init__.py:
```diff
@@ -126,13 +126,43 @@
+def validate_address(register_set: set):
+    def validator(register: dict):
+        address = register[CONF_ADDRESS]
+        register_count = register.get(
+            CONF_REGISTER_COUNT, TYPE_REGISTER_MAP[register[CONF_VALUE_TYPE]]
+        )
```
`CONF_REGISTER_COUNT` is not a valid option in the `ModbusServerRegisterSchema` so this will always default to the second arg

gotnone replied:
I extracted this code from my server_write_registers branch.

I am anticipating needing this check in the upcoming `server_write_registers` PR.  I can remove it, but then I'll just have to add it back in that PR.

I also have an upcoming PR in the works to fix issue https://github.com/esphome/issues/issues/7000.  It will also utilize this.

gotnone replied:
OK removed.

Commit:
[modbus_controller] Use CORE.data for address set

Commit:
[modbus_controller] Fix validate_address

Commit:
[modbus_controller] Fix CORE.data usage

Commit:
[modbus_controller] Use MultipleInvalid.add()

Commit:
[modbus_controller] constant for validate_address
Use constant KEY_SERVER_REGISTERS = "server_registers" in
validate_address() parameter

jesserockz commented on esphome/components/modbus_controller/__init__.py:
```diff
@@ -126,13 +130,42 @@
 
+
+def validate_address(register_set_key: str):
+    register_set = CORE.data.setdefault(DOMAIN, {}).setdefault(register_set_key, set())
+
+    def validator(register: dict):
```
Looking over this again, I think this needs to be inside the actual validator
```suggestion
    def validator(register: dict):
        register_set = CORE.data.setdefault(DOMAIN, {}).setdefault(register_set_key, set())```

gotnone replied:
I had to modify this proposed fix to keep the linter happy, but now it is inside the validator

Commit:
[modbus_controller] register_set into validator

Commit:
[modbus_controller] update to current modbus code

- Merge remote-tracking branch 'origin/dev' into fix_6698

Commit:
[modbus_controller] Fix CODEOWNERS
